pkg/codegen/convertutil/utilities.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convertutil
+
+import (
+	"github.com/blang/semver"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+func SetHighestPackageVersion(versionExpr *model.Expression, pkg string, packages []*schema.Package) {
```

Comment:
Some doc comments on this function to explain what it does would be great.

It looks like it scans through the list of available packages, mutating the `pkg.Version` field if it has a smaller package. This seems scary for multiple reasons:
1. Since loaded packages are shared within the same loader (and often used concurrently while testing), this introduces concurrent global mutable state. 
2. It does not take into account `pluginDownloadURL`. It is entirely possible to have two different packages, both called `pkg` with different `pluginDownloadURL`s. You could increase a packages version such that it references a package that does not exist.

Reply:
You're right, this might not be a safe approach. I think we can collect all seen versions of a package name but not modify the actual package. Then, in `GenerateProgram()` when creating the `requirements.txt`, while iterating over each package we do the check?